Upgrade tool failed to execute

### Node Version

22.14.0

### NPM/Yarn/PNPM Version

PNPM 10.12.1

### Strapi Version

5.15.0

### Operating System

MacOS

### Database

PostgreSQL

### Javascript or Typescript

Typescript

### Reproduction URL

_No response_

### Bug Description

Executing `npx @strapi/upgrade minor` to upgrade the version failed.

### Steps to Reproduce

Executing `npx @strapi/upgrade minor` 

### Expected Behavior

Execute the command to update successfully

### Logs

```shell
[WARN]  [2025-07-09T09:57:59.946Z] Please make sure you've created a backup of your codebase and files before upgrading
[INFO]  [2025-07-09T09:58:01.915Z] Upgrading from v5.15.0 to v5.18.0
[INFO]  [2025-07-09T09:58:01.925Z] (1/4) Checking requirement...
[INFO]  [2025-07-09T09:58:01.962Z] (2/4) Applying the latest code modifications...
[INFO]  [2025-07-09T09:58:01.965Z] (3/4) Upgrading Strapi dependencies...
[INFO]  [2025-07-09T09:58:01.965Z] (4/4) Installing dependencies...
(node:60368) Warning: We detected your package manager (pnpm v>=3), but it's not officially supported by Strapi yet. Defaulting to npm instead.
(Use `node --trace-warnings ...` to show where the warning was created)
npm error Cannot read properties of null (reading 'matches')
npm error A complete log of this run can be found in: xxxxx/2025-07-09T09_58_02_017Z-debug-0.log
[ERROR] [2025-07-09T09:58:12.827Z] Command failed with exit code 1: npm install
```

### Code Snippets

_No response_

### Media

_No response_

### Additional information

_No response_

### Confirmation Checklist

- [x] I have checked the existing [issues](https://github.com/strapi/strapi/issues) for duplicates.
- [x] I agree to follow this project's [Code of Conduct](https://github.com/strapi/strapi/blob/develop/CODE_OF_CONDUCT.md).

> NPM/Yarn/PNPM Version
> PNPM 10.12.1

>(node:67685) Warning: We detected your package manager (pnpm v>=3), but it's not officially supported by Strapi yet. Defaulting to npm instead.

associated with [npm ERR! Cannot read properties of null (reading 'matches')](https://stackoverflow.com/questions/74650198/npm-err-cannot-read-properties-of-null-reading-matches)

> This error shows up when you run npm install in a directory where you previously ran pnpm install.

Is where the issue likely lies

cc @maccomaccomaccomacco & @innerdvations regarding a potential pnpm official support

--- a/packages/core/utils/src/__tests__/package-manager.test.ts
+++ b/packages/core/utils/src/__tests__/package-manager.test.ts
@@ -1,0 +1,52 @@
+import preferredPM from 'preferred-pm';
+import execa from 'execa';
+import * as pkg from '../package-manager';
+
+jest.mock('preferred-pm');
+jest.mock('execa');
+
+const mockedPreferredPM = preferredPM as jest.MockedFunction<typeof preferredPM>;
+const mockedExeca = execa as jest.MockedFunction<typeof execa>;
+
+describe('package-manager', () => {
+  describe('getPreferred', () => {
+    it('returns pnpm when preferred-pm detects pnpm', async () => {
+      mockedPreferredPM.mockResolvedValue({ name: 'pnpm', version: '10.12.1' });
+
+      await expect(pkg.getPreferred('/some/project')).resolves.toBe('pnpm');
+      expect(mockedPreferredPM).toHaveBeenCalledWith('/some/project');
+    });
+
+    it('defaults to npm with a warning when the detected package manager is not supported', async () => {
+      const warnSpy = jest.spyOn(process, 'emitWarning').mockImplementation(() => {});
+      mockedPreferredPM.mockResolvedValue({ name: 'bun', version: '1.2.0' });
+
+      await expect(pkg.getPreferred('/some/project')).resolves.toBe('npm');
+      expect(warnSpy).toHaveBeenCalled();
+      warnSpy.mockRestore();
+    });
+
+    it('throws when preferred-pm cannot detect a package manager', async () => {
+      mockedPreferredPM.mockResolvedValue(undefined);
+
+      await expect(pkg.getPreferred('/some/project')).rejects.toThrow(
+        `Couldn't find a package manager in your project.`
+      );
+    });
+  });
+
+  describe('installDependencies', () => {
+    it('runs pnpm install in the given directory', async () => {
+      mockedExeca.mockResolvedValue({ stdout: '', stderr: '', exitCode: 0 } as Awaited<
+        ReturnType<typeof execa>
+      >);
+
+      await pkg.installDependencies('/project/root', 'pnpm');
+
+      expect(mockedExeca).toHaveBeenCalledWith('pnpm', ['install'], {
+        cwd: '/project/root',
+        stdin: 'ignore',
+      });
+    });
+  });
+});

--- a/packages/core/utils/src/package-manager.ts
+++ b/packages/core/utils/src/package-manager.ts
@@ -3,10 +3,10 @@ import preferredPM from 'preferred-pm';
 
 import type { Options as ProcessOptions } from 'execa';
 
-const SUPPORTED_PACKAGE_MANAGERS = ['npm', 'yarn'];
+const SUPPORTED_PACKAGE_MANAGERS = ['npm', 'yarn', 'pnpm'];
 const DEFAULT_PACKAGE_MANAGER = 'npm' as const;
 
-type SupportedPackageManagerName = 'npm' | 'yarn';
+type SupportedPackageManagerName = 'npm' | 'yarn' | 'pnpm';
 
 export const getPreferred = async (pkgPath: string): Promise<SupportedPackageManagerName> => {
   const pm = await preferredPM(pkgPath);
@@ -16,7 +16,9 @@ export const getPreferred = async (pkgPath: string): Promise<SupportedPackageMan
     throw new Error(`Couldn't find a package manager in your project.`);
   }
 
-  const isPackageManagerSupported = SUPPORTED_PACKAGE_MANAGERS.includes(pm.name);
+  const isPackageManagerSupported = (SUPPORTED_PACKAGE_MANAGERS as readonly string[]).includes(
+    pm.name
+  );
   if (!isPackageManagerSupported) {
     process.emitWarning(
       `We detected your package manager (${pm.name} v${pm.version}), but it's not officially supported by Strapi yet. Defaulting to npm instead.`

--- a/packages/core/utils/src/package-manager.ts
+++ b/packages/core/utils/src/package-manager.ts
@@ -3,10 +3,10 @@ import preferredPM from 'preferred-pm';
 
 import type { Options as ProcessOptions } from 'execa';
 
-const SUPPORTED_PACKAGE_MANAGERS = ['npm', 'yarn', 'pnpm'];
+const SUPPORTED_PACKAGE_MANAGERS = ['npm', 'yarn', 'pnpm'] as const;
 const DEFAULT_PACKAGE_MANAGER = 'npm' as const;
 
-type SupportedPackageManagerName = 'npm' | 'yarn' | 'pnpm';
+type SupportedPackageManagerName = (typeof SUPPORTED_PACKAGE_MANAGERS)[number];
 
 export const getPreferred = async (pkgPath: string): Promise<SupportedPackageManagerName> => {
   const pm = await preferredPM(pkgPath);

--- a/packages/utils/upgrade/src/modules/npm/__tests__/package.test.ts
+++ b/packages/utils/upgrade/src/modules/npm/__tests__/package.test.ts
@@ -250,6 +250,24 @@ describe('Package registry URL determination', () => {
     );
   });
 
+  it('should use pnpm registry when pnpm is the preferred package manager', async () => {
+    const pnpmRegistry = 'https://pnpm-registry.example.com/';
+    mockGetPreferred.mockResolvedValue('pnpm');
+    mockExeca.mockResolvedValue({ stdout: pnpmRegistry } as ExecaReturnValue);
+
+    const pkg = new Package('@test/package', mockCwd, mockLogger);
+    await pkg.refresh();
+
+    expect(mockGetPreferred).toHaveBeenCalledWith(mockCwd);
+    expect(mockExeca).toHaveBeenCalledWith('pnpm', ['config', 'get', 'registry'], {
+      timeout: 10_000,
+    });
+    expect(global.fetch).toHaveBeenCalledWith(
+      'https://pnpm-registry.example.com/@test/package',
+      expect.anything()
+    );
+  });
+
   it('should fallback to default registry when no other registry is available', async () => {
     delete process.env.NPM_REGISTRY_URL;
     mockGetPreferred.mockResolvedValue(null);

--- a/packages/utils/upgrade/src/modules/npm/package.ts
+++ b/packages/utils/upgrade/src/modules/npm/package.ts
@@ -72,6 +72,7 @@ export class Package implements PackageInterface {
       const registryCommands = {
         yarn: ['config', 'get', 'npmRegistryServer'],
         npm: ['config', 'get', 'registry'],
+        pnpm: ['config', 'get', 'registry'],
       } as const;
 
       const command = registryCommands[packageManagerName as keyof typeof registryCommands];

